src/data/variationOfCycleChecker.ts

```diff
@@ -0,0 +1,70 @@
+import {DbTricksTableZod} from "../lib/database/schemas/CurrentVersionSchema"
+import { z } from "zod";
+
+type Trick = z.infer<typeof DbTricksTableZod>
+
+/**
+ * Tricks are set up in a way where variationOf defines a "parent" trick they belong to. 
+ * Important note here is that we do not have trees as a trick can have multiple parents.
+ * 
+ * As such, the set of tricks can be described as a set of disjunct directed cycle-free graphs.
+ * This module is responsible for detecting cycles.
+ *  
+ */
+export default function buildGraph(tricks: Trick[]) {
+
+  /**
+   * This is a "shortcut" map telling us all the roots of a trick. Mainly
+   * here to skip doing redundant work. If a path has already been taken, it can be skipped
+   */
+  const rootsIndex: Record<Trick["id"], Trick["id"][]> = {}
+
+  /**
+   * Create a map of child ID -> parent ID for easy lookup
+   */
+  const childIdToParentIdsLookup: Record<Trick["id"], Trick["id"][]> = {}
+
+  tricks.forEach( e => {
```

Comment:
Maybe this `e` could be renamed to `trick`?

Reply:
done